chpl/chpl-api/src/main/java/gov/healthit/chpl/filter/DeprecatedEndpointUsageFilter.java

```diff
@@ -0,0 +1,151 @@
+import gov.healthit.chpl.util.ApiKeyUtil;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@Component
+public class DeprecatedEndpointUsageFilter extends GenericFilterBean {
+    public static final String[] IGNORED_REQUEST_PATHS = {
+            "/api-docs", "/monitoring", "/ff4j-console"
+    };
+
+    private ApiKeyManager apiKeyManager;
+    private DeprecatedApiUsageDao deprecatedApiUsageDao;
+    private RequestMappingHandlerMapping requestMappingHandlerMapping;
+
+    @Autowired
+    public DeprecatedEndpointUsageFilter(ApiKeyManager apiKeyManager,
+            DeprecatedApiUsageDao deprecatedApiUsageDao,
+            RequestMappingHandlerMapping requestMappingHandlerMapping) {
+        this.apiKeyManager = apiKeyManager;
+        this.deprecatedApiUsageDao = deprecatedApiUsageDao;
+        this.requestMappingHandlerMapping = requestMappingHandlerMapping;
+    }
+
+    @Override
+    public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
+            throws IOException, ServletException {
+        final HttpServletRequest request = (HttpServletRequest) req;
+
+        boolean isIgnored = Stream.of(IGNORED_REQUEST_PATHS).filter(path -> path.matches(request.getServletPath())).findAny().isPresent();
+        if (isIgnored) {
+            chain.doFilter(req, res);
+            return;
+        }
+
+        RequestMappingInfo requestMapping = getRequestMappingForHttpServletRequest(request);
+        if (requestMapping != null) {
+            HandlerMethod handlerMethod = requestMappingHandlerMapping.getHandlerMethods().get(requestMapping);
+            if (isHandlerMethodDeprecated(handlerMethod)) {
+                Set<String> matchingUrlPatterns = requestMapping.getPatternsCondition().getPatterns();
+                if (matchingUrlPatterns != null && matchingUrlPatterns.size() > 0) {
+                    String matchingUrlPattern = matchingUrlPatterns.iterator().next();
+                    LOGGER.warn(request.getRequestURI() + " maps to deprecated endpoint " + matchingUrlPattern + ", handler: " + handlerMethod);
+                    ApiKey apiKey = getApiKey(request);
+                    DeprecatedApi deprecatedApi = getApi(request, matchingUrlPattern);
+                    if (apiKey != null && deprecatedApi != null) {
+                        DeprecatedApiUsage deprecatedApiUsage = DeprecatedApiUsage.builder()
+                                .apiKey(apiKey)
+                                .api(deprecatedApi)
+                                .build();
+                        deprecatedApiUsageDao.createOrUpdateDeprecatedApiUsage(deprecatedApiUsage);
+                    }
+                } else {
+                    LOGGER.error("Could not determine unique matching URL Pattern for " + request.getMethod()
+                        + " Request: " + request.getRequestURI());
+                }
+            }
+        }
+
+        //TODO: something about deprecated parameters
```

Comment:
Todo in a later ticket? Or something for now?

Reply:
Later ticket. Removed